In [7]:
from googleapiclient import discovery
import json
from sklearn.metrics import accuracy_score


API_KEY = 'AIzaSyC0Sqaou4IUy22wHQxuMjb1uc_fzgaAdnY'

client = discovery.build(
  "commentanalyzer",
  "v1alpha1",
  developerKey=API_KEY,
  discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
  static_discovery=False,
)

analyze_request = {
  'comment': { 'text': 'i hate you' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 10, 'score': {'value': 0.68408644, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.68408644, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


Since the API gives scores between 0 and 1. We can set the threshold to 0.5 and consider anything above 0.5 as toxic and below 0.5 as non toxic.


My Hypothesis:
I believe that Perspective will be less able to detect toxicity from comments which incorporate British slang insults rather than common American ones. I think it will make more mistakes with comments in this category.

Test set of statements

American Insults:

1. You're such an asshole.
2. What a complete idiot.
3. That makes you a thot.
4. I don't know why you're a dickhead.
5. This proves you're only meant for blue collar work.

British Insults:

1. You're such a tosser.
2. What a complete wanker.
3. That makes you a slag.
4. I don't know why you're a knob head.
5. This proves you're just a wazzock.

Each British phrase corresponds to the American term used for the same listed number. They're effectively synonyms.

In [24]:

comments_to_analyze = [
    'You\'re such an asshole.',
    'You\'re such a tosser.',
    'What a complete idiot.',
    'What a complete wanker.',
    'That makes you a thot.',
    'That makes you a slag.',
    'I don\'t know why you\'re a dickhead.',
    'I don\'t know why you\'re a knob head',
    'This proves you\'re just a blue collar',
    'This proves you\'re just a wazzock.',

]

for comment_text in comments_to_analyze:
    # Define the analyze request for each comment
    analyze_request = {
        'comment': {'text': comment_text},
        'requestedAttributes': {'TOXICITY': {}}
    }

    # Execute the analyze request
    response = client.comments().analyze(body=analyze_request).execute()

    # Print the response for each comment
    print(f"Comment: {comment_text}")
    print(response)

Comment: You're such an asshole.
{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 23, 'score': {'value': 0.95048624, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.95048624, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}
Comment: You're such a tosser.
{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 21, 'score': {'value': 0.50789946, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.50789946, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}
Comment: What a complete idiot.
{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 22, 'score': {'value': 0.8988238, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.8988238, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}
Comment: What a complete wanker.
{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 23, 'score': {'value': 0.7510937, 'type': 'PROBABILITY'}}], 'summaryScor

1.  "You're such an asshole." This received a toxicity score of 0.95048624. Result: TOXIC

2.  "You're such a tosser." This received a toxicity score of 0.50789946. Result: TOXIC

3. A "What a complete idiot." This received a toxicity score of 0.8988238. Result: TOXIC

4. B "What a complete wanker." This received a toxicity score of 0.7510937. Result: TOXIC

5. "That makes you a thot." This received a toxicity score of 0.50503236. Result: TOXIC

6. "That makes you a slag." This received a toxicity score of 0.37751234. Result: NON TOXIC

7. "I don't know why you're a dickhead." This received a toxicity score of 0.911035. Result: TOXIC

8. "I don't know why you're a knob head." This received a toxicity score of 0.52272606. Result: TOXIC

9. "This proves you're just a blue collar." This received a toxicity score of 0.31963667. Result: NOT TOXIC

10. "This proves you're just a wazzock." This received a toxicity score of 0/4575469. Result: NOT TOXIC

In [25]:
def calculate_accuracy(predictions, actual):
    return accuracy_score(predictions, actual)

def process_and_evaluate(scores, actual, country):
    predictions = ['Yes' if score > 0.5 else 'No' for score in scores]
    accuracy = calculate_accuracy(predictions, actual)
    print(f"Accuracy for {country} data = {accuracy}")

#yes indicates toxic and no indicates not toxic
american_scores = [0.95948624, 0.8988238, 0.50503236, 0.911035, 0.31963667]
american_actual = ['Yes', 'Yes', 'Yes', 'Yes', 'Yes']

british_scores = [0.50789946, 0.7510937, 0.37751234, 0.911035, 0.47575469]
british_actual = ['Yes', 'Yes', 'Yes', 'Yes', 'Yes']

all_scores = british_scores + american_scores
all_actual = british_actual + american_actual


process_and_evaluate(all_scores, all_actual, "all")
process_and_evaluate(british_scores, british_actual, "British")
process_and_evaluate(american_scores, american_actual, "American")

Accuracy for all data = 0.7
Accuracy for British data = 0.6
Accuracy for American data = 0.8


The accuracy for all data was a 0.7. However, the accuracy score for the british data was 0.6 and for American it was 0.8. This shows a slight but very real tendency towards not identifying British slang when toxic.:

My hypothesis was correct based on the samples I utilized. it was more likely to recognize the American terms as toxic and give it a higher rating despite the British terms being synonyms.

Write about Results:

In this, I have learned that models cannot be assumed to account for diverse datasets. It's important to assess the model for any potential bias regarding slang words or terms from different age groups, subcultures, or regions of the world. I assumed that English language content would be the most tested for in this model, but it was clear from this test that American English is but not so much British English. I wasn't surprised from the overall results of these findings, but I didn't expect the difference to be as pronounced as it was.

I theorize that this bias is based on the training data given to the model. I think it was trained on American information, so it struggled to recognize the British slang insults as toxic comments. I think this is a gap in the model

Overall, the average is 0.717002 for the American sentences and 0.5233458 for the British sentences. This is a significant difference. Overall, two British returned as non-toxic and only one American sentence.

We also saw this with the calculated accuracy at the end. It was less accurate for the British statements at 0.6 as compared to the American ones at 0.8.

Interestingly, one of the lowest American sentences included the word "thot" which might be because it's a lesser known slang word more common amoung generation z online users rather than older audiences who worked on the model.